### PR TITLE
Implement EIP-3076 minimal slashing protection, using a filesystem database

### PR DESCRIPTION
**Pull request by nalepae**
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
This pull request implements [EIP-3076 minimal slashing protection](https://www.google.com/search?q=eip+3076+tests&oq=eip+30&gs_lcrp=EgZjaHJvbWUqCAgAEEUYJxg7MggIABBFGCcYOzIGCAEQRRg5MgYIAhBFGEAyCAgDEEUYJxg7MgYIBBBFGDsyBggFEEUYPDIGCAYQRRg8MgYIBxBFGD2oAgCwAgA&sourceid=chrome&ie=UTF-8), using a filesystem database.

**Which issues(s) does this PR fix?**

Fixes #12475 

**Other notes for review**
This PR is designed to be read commit by commit.

This PR implements the flag `--enable-minimal-slashing-protection` in the validator client.
If this flag is used, then the former BoltDB `validator.db` database is not used any more.
Instead, a pure filesystem, YAML database is used, using a EIP-3076 minimal slashing protection database.

The filesystem structure is the following:
```
.
├── configuration.yaml
└── slashing-protection
    ├── 0x80f408a41341416f79e3be605c5f1b960169bea5753b9b8181f7466b4fa0b13addb250c9d068af263f820c113536f9f7.yaml
    ├── 0x84d7fa31925c559ef85531687b190d9299036625a73370153502bf1c086ac7d7e3006c15210c6f9bd4b518c1900e0f45.yaml
    └── 0xa17524bcf06d4fbf042e17dae68637368607f4fc9c28c104f49b12d1e0bed93fa3714f35ff05f4e518f789319c9b8f86.yaml
```

`configuration.yaml` contains data relative to:
- genesis validator root,
- proposer settings, and
- graffiti

An example of `configuration.yaml` file is:
```yaml
genesisValidatorsRoot: 0x043db0d9a83813551ee2f33450d23797757d430911a9320530ad8a0eabc43efb
proposerSettings:
    proposerconfig:
        0x80f408a41341416f79e3be605c5f1b960169bea5753b9b8181f7466b4fa0b13addb250c9d068af263f820c113536f9f7:
            feerecipient: 0xef01d47ed04b181a2715D4050208f7d6e381bF42
            builder:
                enabled: true
                gaslimit: 30000000
                relays: []
        0xa17524bcf06d4fbf042e17dae68637368607f4fc9c28c104f49b12d1e0bed93fa3714f35ff05f4e518f789319c9b8f86:
            feerecipient: 0x801f87A1e14398e2ec9031CaBd585A84D92BC606
            builder:
                enabled: true
                gaslimit: 30000000
                relays: []
    defaultconfig:
        feerecipient: 0x1268AD189526AC0b386faF06eFfC46779c340eE6
        builder:
            enabled: true
            gaslimit: 30000000
            relays: []
graffiti:
    orderedindex: 0
    filehash: "0x0000000000000000000000000000000000000000000000000000000000000000"
```

In the `slashing-protection` directory, each file contains slashing protection data relative a given public key.
The public key itself is the name of the file.

An example of a file in the `slashing-protection` directory is:
```yaml
latestSignedBlockSlot: 224123
lastSignedAttestationSourceEpoch: 224990
lastSignedAttestationTargetEpoch: 224991
```

The content of each field is self-describing.

Regarding anti-slashing strategy:
- For block proposals: The signature of an incoming block proposal will be refused if the slot of the incoming block is **lower or equal to** the `latestSignedBlockSlot` stored in the database.
- For attestations: The signature of an incoming attestation will be refused if:
   - the source epoch of the incoming attestation is **lower to** the `lastSignedAttestationSourceEpoch` stored in the database, or if
   - the target epoch of the incoming attestation is **lower or equal to** the `lastSignedAttestationSourceEpoch` stored in the database.

**Strategy at VC start:**
- If both minimal, filesystem AND complete, BoltDB databases are found, the VC refuses to start
- If `--enable-minimal-slashing-protection` is set AND a complete, BoltDB database is found, then a warning message is displayed indicating to the user how to convert the DB, and the VC continues with the complete, BoltDB database.
- If `--enable-minimal-slashing-protection` is NOT set AND a minimal, filesystem database is found, then the VC automatically convert the minimal, filesystem database into a complete, BoltDB one. (No data loss in this way)

To convert a complete database into a minimal one, the user can do:
```
validator db convert-complete-to-minimal [--datadir <datadir>]
```
**😱 This PR has 10k+ new lines! 😱**
Actualy, a lot of modified lines in this PR result in tests transformed from something like:
```go
func test(t *testing.T) {
   ... := setup(t)
   // Content of the test
}
```
to:
```go
func test(t *testing.T) {
   for _, isSlashingProtectionMinimal := range [...]bool{false, true} {
      ... := setup(t, isSlashingProtectionMinimal)
      // Content of the test
}
```

Git will consider all wrapped lines as new lines, even if only a few spaces had been added on the left.
For this reason, it is strongly advised to check the `Hide whitespace` option during the code review:
image

**Benchmarks:**
We create `2000` public keys, and we call, in parallel, `SaveAttestationForPubKey` on both complete, BoltDB and minimal, filesystem DBs.

Benchmark results:
- complete, BoltDB: `~130ms`
- minimal, filesystem: `~127ms`

Comments:
For complete, BoltDB database, `SaveAttestationForPubKey` fills a queue which is dumped into the DB every 100 ms (or if the queue, containing up to `2048` items, is full). It explains why the time needed is `>= 100ms`. If we decide to save only `10` attestations instead of `2000`, the benchmark result is still `~104ms`. Obviously, it cannot go under `100ms`.

### CODE REVIEW

File: validator/db/filesystem/proposer_settings.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+		// Save the configuration.
+		if err := s.saveConfiguration(configuration); err != nil {
+			return errors.Wrap(err, "could not save configuration")
+		}
+
+		return nil
+	}
+
+	if configuration.ProposerSettings == nil {
```
Comment (rauljordan) [inline validator/db/filesystem/proposer_settings.go:99]: Could be all done in the if above with configuration == nil || configuration.ProposerSettings == nil

File: validator/db/filesystem/db.go [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+	configurationFileName     = "configuration.yaml"
+
+	DatabaseDirName = "validatorClientData"
+)
+
+type (
+	// Store is a filesystem implementation of the validator client database.
+	Store struct {
+		configurationMu    sync.RWMutex
+		pkToSlashingMu     map[[fieldparams.BLSPubkeyLength]byte]*sync.RWMutex
```
Comment (rauljordan) [inline validator/db/filesystem/db.go:36]: Mutexes in go are designed to be used without pointers, as using pointers can lead to issues. If this gets copied, for example, you will have two pointers to the same mutex. The default value of a mutex is also valid, so it should be safe to keep it as such

File: validator/db/filesystem/db.go [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+	// Get all entries in the slashing protection directory.
+	entries, err := os.ReadDir(slashingProtectionDirPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not read database directory")
+	}
+
+	// Collect public keys.
+	publicKeys := make([][fieldparams.BLSPubkeyLength]byte, 0, len(entries))
+	for _, entry := range entries {
+		if !(entry.Type().IsRegular() && strings.HasPrefix(entry.Name(), "0x")) {
```
Comment (rauljordan) [inline validator/db/filesystem/db.go:381]: Must they be 0x prefixed? Or is it optional in the spec? My worries here is that we just continue and don't log anything to the user. Perhaps we should offer some info that something was ignored and raise a warning

**Review by rauljordan — COMMENTED** [review]
Very comprehensive, complete feature implementation. My concerns of the big line count were reduced by seeing the quality of tests and comments. Something to keep in mind is this feature will also need careful docs and communications to users on how to start employing it. Following a rigorous checklist for bringing it to prod can help

File: cmd/validator/flags/flags.go [hunk 4]
```diff
@@ -377,6 +377,20 @@ var (
 		Usage: "Sets the maximum size for one batch of validator registrations. Use a non-positive value to disable batching.",
 		Value: 0,
 	}
+
+	// SourceDataDirFlag defines a path on disk where source Prysm databases are stored. Used for conversion.
+	SourceDataDirFlag = &cli.StringFlag{
+		Name:     "source",
```
[the remark this answers is not in the document]
Comment (nalepae) [reply cmd/validator/flags/flags.go:383]: Definitely!

File: cmd/validator/flags/flags.go [hunk 5]
```diff
@@ -377,6 +377,20 @@ var (
 		Usage: "Sets the maximum size for one batch of validator registrations. Use a non-positive value to disable batching.",
 		Value: 0,
 	}
+
+	// SourceDataDirFlag defines a path on disk where source Prysm databases are stored. Used for conversion.
+	SourceDataDirFlag = &cli.StringFlag{
+		Name:     "source-data-dir",
+		Usage:    "Source data directory",
```
Comment (james-prysm) [inline cmd/validator/flags/flags.go:384]: can you use a more descriptive version for the flag usage? https://docs.prylabs.network/docs/prysm-usage/parameters it shows up on this for users

Comment (james-prysm) [reply cmd/validator/flags/flags.go:384]: you also need to add it here i believe https://github.com/prysmaticlabs/prysm/blob/develop/cmd/validator/usage.go

File: validator/client/BUILD.bazel [hunk 6]
```diff
@@ -97,10 +96,9 @@ go_library(
 
 go_test(
     name = "go_default_test",
-    size = "small",
+    size = "medium",
```
Comment (james-prysm) [inline validator/client/BUILD.bazel:99]: why the change here?

File: validator/client/attest.go [hunk 7]
```diff
@@ -27,6 +27,8 @@ import (
 	"go.opencensus.io/trace"
 )
 
+var failedAttLocalProtectionErr = "attempted to make slashable attestation, rejected by local slashing protection"
```
Comment (james-prysm) [inline validator/client/attest.go:30]: could be a const

File: validator/client/propose_protect.go [hunk 8]
```diff
@@ -3,20 +3,45 @@ package client
...
 
 // slashableProposalCheck checks if a block proposal is slashable by comparing it with the
 // block proposals history for the given public key in our DB. If it is not, we then update the history
 // with new values and save it to the database.
 func (v *validator) slashableProposalCheck(
-	ctx context.Context, pubKey [fieldparams.BLSPubkeyLength]byte, signedBlock interfaces.ReadOnlySignedBeaconBlock, signingRoot [32]byte,
+	ctx context.Context,
+	pubKey [fieldparams.BLSPubkeyLength]byte,
+	signedBlock interfaces.ReadOnlySignedBeaconBlock,
+	signingRoot [32]byte,
```
Comment (james-prysm) [inline validator/client/propose_protect.go:25]: you can also use fieldparams.RootLength

File: validator/db/filesystem/proposer_settings.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+		return nil, errors.Wrap(err, "could not get configuration")
+	}
+
+	// Return on error if config file does not exist.
+	if configuration == nil || configuration.ProposerSettings == nil {
+		return nil, ErrNoProposerSettingsFound
+	}
+
+	// Convert proposer settings to validator service config.
+	proposerSettings, err := validatorServiceConfig.ToSettings(configuration.ProposerSettings)
```
Comment (james-prysm) [inline validator/db/filesystem/proposer_settings.go:31]: hmm i think this should still work fine with changes I'm proposing

File: validator/db/filesystem/proposer_settings.go [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	validatorServiceConfig "github.com/prysmaticlabs/prysm/v4/config/validator/service"
+	validatorpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1/validator-client"
+)
+
+// ErrNoProposerSettingsFound is an error thrown when no settings are found.
+var ErrNoProposerSettingsFound = errors.New("no proposer settings found in bucket")
+
+// ProposerSettings returns the proposer settings.
+func (s *Store) ProposerSettings(_ context.Context) (*validatorServiceConfig.ProposerSettings, error) {
```
Comment (james-prysm) [inline validator/db/filesystem/proposer_settings.go:18]: is it worth using the context here for a trace log?

File: validator/node/node_test.go [hunk 11]
```diff
@@ -849,80 +853,86 @@ func TestProposerSettings(t *testing.T) {
...
+			set.Bool(flags.EnableBuilderFlag.Name, true, "")
+			cliCtx := cli.NewContext(&app, set, nil)
+			validatorDB := dbTest.SetupDB(t, [][fieldparams.BLSPubkeyLength]byte{}, isSlashingProtectionMinimal)
+			got, err := proposerSettings(cliCtx, validatorDB)
+			require.NoError(t, err)
+			_, err = validatorDB.ProposerSettings(cliCtx.Context)
+			require.ErrorContains(t, "no proposer settings found in bucket", err)
+			want := &validatorserviceconfig.ProposerSettings{
+				DefaultConfig: &validatorserviceconfig.ProposerOption{
+					BuilderConfig: &validatorserviceconfig.BuilderConfig{
```
Comment (james-prysm) [inline validator/node/node_test.go:928]: my changes affect this test

File: io/file/fileutil.go [hunk 12]
```diff
@@ -183,6 +183,40 @@ func RecursiveFileFind(filename, dir string) (bool, string, error) {
 	return found, fpath, nil
 }
 
+// RecursiveDirFind searches for directory in a directory and its subdirectories.
+func RecursiveDirFind(dirname, dir string) (bool, string, error) {
```
Comment (james-prysm) [inline io/file/fileutil.go:187]: maybe I should use this for the keystore import, mine only goes 2 levels deep...

File: validator/client/propose_protect.go [hunk 13]
```diff
@@ -3,20 +3,45 @@ package client
...
+) error {
+	switch v.db.(type) {
```
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/client/propose_protect.go]: Fixed, I remove all type switches on db. For attestations, proposals, and imports.
Now export is "alone", I guess to be consistent I would need to move it as well.
But this PR is already very big, let's reserve it for a next one.

File: validator/db/filesystem/db.go (hunk 2 above)
Comment (nalepae) [reply validator/db/filesystem/db.go:37]: Maybe I missed something, but I need to mutate mutexes in the map (`mu.RLock()` and `mu.RUnlock()`).

https://github.com/uber-go/guide/blob/master/style.md#receivers-and-interfaces

File: validator/db/convert.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+	"github.com/prysmaticlabs/prysm/v4/validator/db/iface"
+	"github.com/prysmaticlabs/prysm/v4/validator/db/kv"
+)
+
+// IsCompleteDatabaseExisting checks if a complete database exists in the given data directory
+func IsCompleteDatabaseExisting(dataDir string) (bool, error) {
+	// Construct the path to the complete database.
+	completeDatabasePath := filepath.Join(dataDir, kv.ProtectionDbFileName)
+
+	if _, err := os.Stat(completeDatabasePath); err == nil {
```
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/db/convert.go]: The `file.Exists` utility exists, but:
- Does detect only regular files, not directories, and
- Returns `false` on error.

==> I updated the utility to solve these 2 issues and I use it.

File: validator/db/convert.go [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+		// If the error is not a "does not exist" error, return the error.
+		return false, errors.Wrapf(err, "could not stat file %s", completeDatabasePath)
+	}
+
+	// The only error that remains is a "does not exist" error.
+	return false, nil
+}
+
+// IsMinimalDatabaseExisting checks if a minimal database exists in the given data directory
+func IsMinimalDatabaseExisting(dataDir string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/db/convert.go:31]: Actually there is an issue in `IsMinimalDatabaseExisting`, the logic is actually duplicated but should not be.
Fixed.

File: validator/db/convert.go [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/validator/db/filesystem"
+	"github.com/prysmaticlabs/prysm/v4/validator/db/iface"
+	"github.com/prysmaticlabs/prysm/v4/validator/db/kv"
+)
+
+// IsCompleteDatabaseExisting checks if a complete database exists in the given data directory
+func IsCompleteDatabaseExisting(dataDir string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/db/convert.go:18]: Fixed: I deleted those functions.

File: validator/db/filesystem/proposer_settings.go (hunk 1 above)
Comment (nalepae) [reply validator/db/filesystem/proposer_settings.go:99]: - If `configuration == nil`, then we create a brand new configuration file.
- If `configuration.ProposerSettings`, then we use the currently existing configuration, and only modify the `ProposerSettings` part.

I don't really see how I could do with only a `if configuration == nil || configuration.ProposerSettings == nil`.

File: validator/db/filesystem/db.go (hunk 3 above)
Comment (nalepae) [reply validator/db/filesystem/db.go:405]: There is no spec at all actually for how the data should be stored.
The `0x` prefix is only here to filter potential trash files a user would put in the directory.

> Perhaps we should offer some info that something was ignored and raise a warning.
Definitely, added.

File: cmd/validator/flags/flags.go (hunk 5 above)
Comment (nalepae) [reply cmd/validator/flags/flags.go:384]: `SourceDataDirFlag` and `TargetDataDirFlag` do not show on `validator --help`, because they are not registered in `usage.go`.

These options are only used (and advertised) when using `validator db convert-complete-to-minimal`.

==> I moved these options directly in `cmd/validator/db/db.go`.

File: validator/client/BUILD.bazel (hunk 6 above)
Comment (nalepae) [reply validator/client/BUILD.bazel:97]: Because, with new added tests,  if `small`:
```
//validator/client:go_default_test                                      TIMEOUT in 60.0s
```

Tests are longer to run mainly because some tests now run for both complete, boltDB and minimal, filesystem databases.

File: validator/client/attest.go (hunk 7 above)
Comment (nalepae) [reply validator/client/attest.go:30]: Fixed.

File: validator/client/propose_protect.go (hunk 8 above)
Comment (nalepae) [reply validator/client/propose_protect.go]: Fixed for this occurrence and all other `signingRoot` occurrences as well.

File: validator/db/filesystem/proposer_settings.go (hunk 10 above)
Comment (nalepae) [reply validator/db/filesystem/proposer_settings.go:18]: I don't use the context at all for the filesystem DB.
However, context is historically defined in the interface for the BoltDB database. It is effectively used for a trace log.

File: cmd/validator/flags/flags.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nalepae) [reply cmd/validator/flags/flags.go]: Fixed here: https://github.com/prysmaticlabs/prysm/pull/13360/commits/ffa28bb8e01f419677c2e8e7b52167e8d4dbedd7.

File: validator/client/propose_protect.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/client/propose_protect.go]: Fixed with commits from https://github.com/prysmaticlabs/prysm/pull/13360/commits/3a7c10598283fc9b1d26f802881e08a3ccb93810 to https://github.com/prysmaticlabs/prysm/pull/13360/commits/be3d13c9917037e894011679726ac8ea6148909e (included).

File: validator/db/convert.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/db/convert.go]: Fixed with commits from https://github.com/prysmaticlabs/prysm/pull/13360/commits/65e94f46a2a39ccfc2b059727bf1d73dd72066b6 to https://github.com/prysmaticlabs/prysm/pull/13360/commits/59d13c9c789d036162a255ebb5fa44bea22d4884 (included).

File: validator/db/convert.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/db/convert.go]: Fixed in https://github.com/prysmaticlabs/prysm/pull/13360/commits/ab2be4e4ef2543db6c1b6d9f5302b0e007eef25b.

File: validator/db/convert.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/db/convert.go]: Fixed in https://github.com/prysmaticlabs/prysm/pull/13360/commits/ab2be4e4ef2543db6c1b6d9f5302b0e007eef25b.

File: validator/db/convert.go [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+		return false, errors.Wrapf(err, "could not stat file %s", minimalDatabaseConfigurationFilePath)
+	}
+
+	// The only error that remains is a "does not exist" error.
+	return false, nil
+}
+
+// ConvertDatabase converts a minimal database to a complete database or a complete database to a minimal database.
+// Delete the source database after conversion.
+func ConvertDatabase(ctx context.Context, sourceDataDir string, targetDataDir string, minimalToComplete bool) error {
```
[the remark this answers is not in the document]
Comment (nalepae) [reply validator/db/convert.go:20]: Added progress bars when needed (not only for `convert`) in this commit: https://github.com/prysmaticlabs/prysm/pull/13360/commits/0d907084b1c07aa500f04aa9cf9fb795dbff1efb.

**Review by rauljordan — APPROVED** [review]
Leaving approval, thanks for the nice commit diffs! Before we ship, it would still be ideal to measure how long it will actually take to migrate a big DB to the new format just for sanity

**Comment by nalepae** [comment]
> Leaving approval, thanks for the nice commit diffs! Before we ship, it would still be ideal to measure how long it will actually take to migrate a big DB to the new format just for sanity

Converting a complete DB to a minimal one with:
- `2000` validators, and
- `82.125` attestations for each validator (1 year)
takes `12 seconds` on my computer. (Mac M2)

```
INFO: Running command line: bazel-bin/cmd/validator/validator_/validator db convert-complete-to-minimal --source-data-dir /Users/manu/Library/Eth2 --target-data-dir /Users/manu/Library/Eth2
Processing attestations: 100% [======================================]  [12s:0s]
```

**Review by rauljordan — APPROVED** [review]
